cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java

```diff
@@ -92,13 +105,30 @@ void mergeToken(BasicWorkflowToken other) {
 
   @Override
   public void put(String key, String value) {
-    if (nodeName == null) {
-      throw new IllegalStateException("Node name cannot be null.");
+    put(key, Value.of(value));
+  }
+
+  @Override
+  public void put(String key, Value value) {
+    put(key, value, Scope.USER);
+  }
+
+  void put(String key, Value value, Scope scope) {
+    Preconditions.checkNotNull(key, "Null key cannot be added in the WorkflowToken.");
+    Preconditions.checkNotNull(value, String.format("Null value provided for the key '%s'.", key));
+    Preconditions.checkNotNull(value.toString(), String.format("Null value provided for the key '%s'.", key));
+    Preconditions.checkState(nodeName != null);
+
+    Map<String, List<NodeValueEntry>> tokenValuesWithinScope = tokenValueMap.get(scope);
+    if (tokenValuesWithinScope == null) {
+      tokenValuesWithinScope = Maps.newHashMap();
```

Comment:
However, will the code look cleaner if you initialize ``tokenValueMap`` for both ``Scope.USER`` and ``SCOPE.SYSTEM`` in the constructor itself? That way, you won't have to do this null check here and perhaps at other places too.